Implement System.IO.Pipes for Unix

stephentoub opened the pull request:
This PR provides an initial implementation of the System.IO.Pipes library on Unix.

A few notes:
- Ignore almost all of the changes in the *.Windows.cs files... I just moved some code around to put code above or below a "PAL Layer" comment, as we've done in other places in CoreFX.  The only interesting change in the Windows code was that in the named pipe client, I refactored out the retry/cancellation/timing loop so that it could be used by the Unix code as well, and in doing so I added some slightly smarter spinning support.  There were also one or two places where some validation logic in shared code was overly aggressive in what it was checking for when it came to Unix, so I refactored those checks into a core piece and a Windows-specific additional check.
- The .NET APIs map fairly closely to the Windows APIs, and thus there will be some rough edges here on the Unix side, e.g. calling WaitForConnection with named pipes and an InOut direction won't block as it does on Windows.
- I use a bit of "light up" to make some things work on Linux that probably won't as-is on Mac, e.g. in anonymous pipes, I mapped non-inheritability to using O_CLOEXEC, but pipe2 which accepts a flags argument that allows O_CLOEXEC to be specified is Linux-specific (as is dup3 that would allow for reopening a handle with O_CLOEXEC).  This doesn't seem particularly problematic, though, since the primary use case for anonymous pipes is having them be inherited... just another of the rough edges.
- As part of this, I also realized that in the various places in CoreFX we were making libc write calls we might have been losing data, as write doesn't guarantee it'll write out all of the supplied data. I fixed those call sites.

stephentoub commented:
cc: @ellismg, @jaredpar, @pgavlin

Fixes #809

stephentoub commented:
@dotnet-bot test this please

stephentoub commented:
(I've removed the strerror changes; we can address fixing it in a separate PR.)

jaredpar commented on src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs:
```diff
@@ -16,19 +13,29 @@ namespace System.IO.Pipes
+            {
+                // Open the file.  For In or Out, this will block until a client has connected.
+                // Unfortunately for InOut it won't; this is different from the Windows behavior;
+                // it won't block for InOut until it actually performs a read or write operation.
```
It's hard to know at a glance to which OS this last line refers. 

First time I've seen a double `;` ;)

stephentoub replied:
Heh, yeah, not exactly stellar grammar.  I'll fix, thanks.

jaredpar commented on src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Windows.cs:
```diff
@@ -42,100 +32,67 @@ private void ConnectInternal(int timeout, CancellationToken cancellationToken, i
-                    waitTime = Math.Min(CancellationCheckInterval, timeLeft);
-                }
-                else
+                // Server is not yet created so let's keep looping.
```
This function doesn't loop anymore.

stephentoub replied:
Yeah, the loop got moved out to the caller.  Will fix the comment.  Thanks.

jaredpar commented on src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs:
```diff
@@ -126,11 +122,138 @@ public Task WaitForConnectionAsync(CancellationToken cancellationToken)
+        private unsafe static readonly IOCompletionCallback s_WaitForConnectionCallback =
+            new IOCompletionCallback(NamedPipeServerStream.AsyncWaitForConnectionCallback);
+
+#if RunAs
```
What is `RunAs` and why isn't it used as `RUNAS`.  The latter seems to be the more common way of defining preprocessor symbols

stephentoub replied:
I actually don't know... I didn't muck with this code, just moved it around in the file. (That's true for almost all of the code in the *.Windows.cs files.)

pgavlin commented on src/System.Console/src/System/ConsolePal.Unix.cs:
```diff
@@ -267,15 +267,17 @@ private static unsafe int Read(int fd, byte[] buffer, int offset, int count)
-                return (int)result;
+                while (count > 0)
+                {
+                    int result;
```
Maybe replace "result" with "written" or "bytesWritten"

stephentoub replied:
Yup, will fix.

pgavlin commented on src/System.IO.Pipes/src/System/IO/Pipes/AnonymousPipeServerStream.Unix.cs:
```diff
@@ -18,9 +17,56 @@ public sealed partial class AnonymousPipeServerStream : PipeStream
+                            while (Interop.CheckIo(Interop.libc.pipe2(fdsptr, (int)Interop.libc.OpenFlags.O_CLOEXEC))) ;
+                            created = true;
+                        }
+                        catch (MissingMethodException) { } // pipe2 may not exist
```
This feels questionable. Why not just PAL this out differently for Linux vs. POSIX?

stephentoub replied:
I agree it's questionable, but I went this way because having an entirely separate binary just to be able to handle one enum value option felt like overkill.  We do similar "light-up" in other places in CoreFX, such as in the metadata reader, and this seemed like an equivalent case.

pgavlin replied:
Cam you define "separate binary" in this case? Do you mean a separate build or a separate library?

stephentoub replied:
Separate build / another .dll that must be distributed in a NuGet package.

pgavlin commented on src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs:
```diff
@@ -4,21 +4,42 @@
 {
     public abstract partial class PipeStream : Stream
     {
+        private const string PipeDirectoryPath = "/tmp/";
```
Probably don't want to put these directly under tmp. Perhaps "/tmp/corefxpipes" would be better?

stephentoub replied:
Ok.

pgavlin commented on src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs:
```diff
@@ -4,21 +4,42 @@
     {
+        private const string PipeDirectoryPath = "/tmp/";
+
+        internal static string GetPipePath(string serverName, string pipeName)
```
Is there any sort of validation performed on this argument by this method's caller?  It would be good to ensure that this is literally a name and not a path.

stephentoub replied:
You mean serverName?  Since we don't support cross-machine pipes in this implementation, I check whether serverName is either "." or the value returned by gethostname, and if it's neither, it throws a PlatformNotSupportedException.

pgavlin replied:
No, I meant pipeName.

stephentoub replied:
I don't explicitly check to see whether pipeName contains any directory separator characters, but such usage should be caught by open and cause a failure.  I can add an explicit check for it.

pgavlin replied:
I think this should be validated here. Neither open() nor mkfifo() cares if the name contains directory separators.

stephentoub replied:
I meant that mkfifo/open would fail because they'd end up looking for a file in a directory that didn't exist; there are certainly ways you could construct a path to get around that, though, e.g. with "../" to escape the tmp directory and then navigate elsewhere.  I've added the validation.

pgavlin replied:
Great. It's exactly those cases that I was thinking of. :)

stephentoub commented:
Jared and Pat, thanks for the feedback!  I addressed it all yesterday, and I just rebased to pick up conflicting changes.  Since this has already been open for a few days and feedback has been relatively minimal, I'm going to go ahead and merge.  Any additional feedback is welcome, and I'll just address it as part of a separate PR.

jaredpar commented:
:+1:

stephentoub commented:
The "netfx-port-consider" tag is specifically for the use of SpinWait, related to https://github.com/dotnet/roslyn/issues/10413
